src/pgen/read_vtk.cpp

```diff
@@ -0,0 +1,900 @@
+
+void Mesh::InitUserMeshData(ParameterInput *pin) {
+  EnrollUserBoundaryFunction(BoundaryFace::inner_x1, SixRayBoundaryInnerX1);
+  EnrollUserBoundaryFunction(BoundaryFace::outer_x1, SixRayBoundaryOuterX1);
+  EnrollUserBoundaryFunction(BoundaryFace::inner_x2, SixRayBoundaryInnerX2);
+  EnrollUserBoundaryFunction(BoundaryFace::outer_x2, SixRayBoundaryOuterX2);
+  EnrollUserBoundaryFunction(BoundaryFace::inner_x3, SixRayBoundaryInnerX3);
+  EnrollUserBoundaryFunction(BoundaryFace::outer_x3, SixRayBoundaryOuterX3);
+  G0 = pin->GetOrAddReal("chem_radiation", "G0", 0.);
+  G0_iang.NewAthenaArray(6);
+  G0_iang(BoundaryFace::inner_x1) = pin->GetOrAddReal("chem_radiation","G0_inner_x1",G0);
+  G0_iang(BoundaryFace::inner_x2) = pin->GetOrAddReal("chem_radiation","G0_inner_x2",G0);
+  G0_iang(BoundaryFace::inner_x3) = pin->GetOrAddReal("chem_radiation","G0_inner_x3",G0);
+  G0_iang(BoundaryFace::outer_x1) = pin->GetOrAddReal("chem_radiation","G0_outer_x1",G0);
+  G0_iang(BoundaryFace::outer_x2) = pin->GetOrAddReal("chem_radiation","G0_outer_x2",G0);
+  G0_iang(BoundaryFace::outer_x3) = pin->GetOrAddReal("chem_radiation","G0_outer_x3",G0);
+  cr_rate = pin->GetOrAddReal("chem_radiation", "CR", 2e-16);
+  return;
+}
+
+//======================================================================================
+//! \fn void MeshBlock::ProblemGenerator(ParameterInput *pin)
+//! \brief initialize problem by reading in vtk file.
+//======================================================================================
+void MeshBlock::ProblemGenerator(ParameterInput *pin) {
+  //dimensions of meshblock
+  const int Nx = ie - is + 1;
+  const int Ny = je - js + 1;
+  const int Nz = ke - ks + 1;
+  //dimensions of mesh
+  const int Nx_mesh = pmy_mesh->mesh_size.nx1;
+  const int Ny_mesh = pmy_mesh->mesh_size.nx2;
+  const int Nz_mesh = pmy_mesh->mesh_size.nx3;
+
+  //read joined or unjoined vtk files
+  int isjoinedvtk = pin->GetOrAddInteger("problem", "is_joined_vtk", 0);
+  AthenaArray<Real> data; //temporary array to store data;
+  AthenaArray<Real> b; //needed for PrimitiveToConserved()
+  if (isjoinedvtk != 0) {
+#ifdef DEBUG
+    printf("Joined vtk file. data size = (%d, %d, %d)\n",
+           Nz_mesh, Ny_mesh, Nx_mesh);
+#endif
+    data.NewAthenaArray(Nz_mesh, Ny_mesh, Nx_mesh);
+    b.NewAthenaArray(Nz_mesh, Ny_mesh, Nx_mesh);
+  } else {
+#ifdef DEBUG
+    printf("Unjoined vtk file. data size = (%d, %d, %d)\n",
+           Nz, Ny, Nx);
+#endif
+    data.NewAthenaArray(Nz, Ny, Nx);
+    b.NewAthenaArray(Nz, Ny, Nx);
+  }
+  std::stringstream msg; // error message
+  std::string vtkfile; // corresponding vtk file for this meshblock
+  // initial abundance
+  const Real r_init = pin->GetReal("problem", "r_init");
+
+  // parse input parameters
+  std::string vtkfile0 = pin->GetString("problem", "vtkfile");//id0 file
+  std::string str_scalers = pin->GetString("problem", "scalers");
```

Comment:
these should all be renamed "scaler" ---> "scalar"

Reply:
Thank you! Fixed.